draft-ietf-ppm-dap.md

```diff
@@ -2587,6 +2595,62 @@ the task and all data pertaining to this task after `task_expiration`.
 Implementors SHOULD provide for some leeway so the Collector can collect the
 batch after some delay.
 
+### Distributed-systems and Synchronization Concerns
+
+Various parts of a DAP implementation will need to synchronize in order to
+ensure correctness during concurrent operation. This section describes the
+relevant concerns and makes suggestions as to potential implementation
+tradeoffs.
+
+* The upload interaction requires the Leader to ignore uploaded reports with a
+  duplicated ID, including concurrently-uploaded reports. This might be
+  implemented by synchronization or via an eventually-consistent process. If the
+  Leader wishes to alert the Client with a `reportRejected` error,
+  synchronization will be necessary to ensure all but one concurrent request
+  receive the error.
+
+* The Leader is responsible for generating aggregation jobs, and will generally
+  want to place each report in exactly one aggregation job. This may require
+  synchronization between different components of the system which are
+  generating aggregation jobs. Note that placing a report into more than one
+  aggregation job will result in a loss of throughput, rather than a loss of
+  correctness, privacy, or robustness, so it is acceptable for implementations
+  to use an eventually-consistent scheme which may rarely place a report into
+  multiple aggregation jobs.
+
+* Aggregation is implemented as a sequence of aggregation steps by both the
+  Leader and the Helper. The Leader must ensure that each aggregation job is
+  only processed once concurrently, which may require synchronization between
+  the components responsible for performing aggregation. The Helper must ensure
+  that concurrent requests against the same aggregation job are handled
+  appropriately, which requires synchronization between the components handling
+  aggregation requests.
+
+* Aggregation requires checking and updating used-report storage as part of
+  implementing replay protection. This must be done while processing the
+  aggregation job, though which steps the checks are performed at is up to the
+  implementation. The checks and storage require synchronization, so that if two
+  aggregation jobs contianing the same report are processed, at most one
+  instance of the report will be aggregated. However, the interaction with the
+  used-report storage does not necessarily have to be synchronized with the
+  processing and storage for the remainder of the aggregation process -- for
+  example, used-report storage could be implemented in a separate datastore than
+  is used for the remainder of data storage, without any transactionality
+  between updates to the two datastores.
+
+* The aggregation and collection interactions require synchronization to avoid
```

Comment:
I believe we point this out elsewhere in operational considerations. If so, perhaps we should remove that text and replace with a reference to this section.

Reply:
I thought so too, but I can't find any text to this effect. The closest I can find is some text in an implementation note in the Input Share Validation section, though it doesn't mention synchronization explicitly. I added a reference to this section to that implementation note.